File: app/src/main/java/com/hyc/qdaily/base/BaseFragment.java
package com.hyc.qdaily.base;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import io.reactivex.Observable;

/**
 * Created by ray on 16/8/31.
 */
public abstract class BaseFragment<T extends BasePresenter> extends Fragment {
    protected T mPresenter;

    protected abstract void initPresenter();


    @Override public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initPresenter();
    }
}

File: app/src/main/java/com/hyc/qdaily/base/BaseFragment.kt
package com.hyc.qdaily.base

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import butterknife.Unbinder

/**
 * Created by ray on 16/8/31.
 */
abstract class BaseFragment<T : BasePresenter<*>> : Fragment() {
    protected var mPresenter: T? = null
    protected var mUnbinder: Unbinder? = null
    protected abstract fun initPresenter()
    protected abstract fun getLayoutID(): Int
    protected abstract fun initView()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initPresenter()
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        var view = inflater!!.inflate(getLayoutID(), container, false)
        mUnbinder = ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
    }

    override fun onDetach() {
        super.onDetach()
        mUnbinder?.unbind()
    }
}
